File: benchmarks/fixtures/webapp_rb/tasks/payment_task.rb
# Payment background tasks.

require_relative '../utils/helpers'
require_relative '../database/connection'
require_relative '../database/queries'
require_relative '../services/payment/processor'
require_relative '../events/dispatcher'

# Process pending payments.
def process_pending_payments(db, events)
  get_logger('tasks.payment').info('Processing pending payments')
  queries = PaymentQueries.new(db)
  pending = queries.find_user_payments('', 'pending')
  processed = 0
  failed = 0
  pending.each do |payment|
    begin
      queries.update_status(payment[:transaction_id], 'completed')
      processed += 1
    rescue StandardError
      queries.update_status(payment[:transaction_id], 'failed')
      failed += 1
    end
  end
  { processed: processed, failed: failed }
end

# Reconcile payments.
def reconcile_payments(db, events)
  get_logger('tasks.payment').info('Reconciling payments')
  queries = PaymentQueries.new(db)
  processing = queries.find_user_payments('', 'processing')
  resolved = 0
  processing.each do |payment|
    queries.update_status(payment[:transaction_id], 'completed')
    resolved += 1
  end
  { resolved: resolved }
end

File: benchmarks/fixtures/webapp_rs/tasks/payment_task.rs
use crate::utils::helpers::get_logger;
use crate::services::payment::processor::PaymentProcessor;
use crate::services::payment::gateway::PaymentGateway;
use crate::services::base::Service;

/// A background task for processing pending payments.
pub struct PaymentTask {
    /// The payment processor instance.
    processor: PaymentProcessor,
    /// Number of payments processed.
    processed: u64,
}

impl PaymentTask {
    /// Create a new payment task with the given API key.
    pub fn new(api_key: &str) -> Self {
        let logger = get_logger("tasks.payment");
        logger.info("Creating PaymentTask");
        let gateway = PaymentGateway::new(api_key, "sandbox");
        Self {
            processor: PaymentProcessor::new(gateway),
            processed: 0,
        }
    }

    /// Run the payment task, processing pending payments.
    pub fn run(&mut self) -> Result<u64, String> {
        let logger = get_logger("tasks.payment");
        self.processor.initialize()?;
        logger.info("Running payment task");
        match self.processor.process_payment("user-1", 99.99, "USD", "card_test") {
            Ok(result) => {
                self.processed += 1;
                logger.info(&format!("Payment processed: txn={}", result.transaction_id));
            }
            Err(e) => {
                logger.error(&format!("Payment failed: {}", e));
            }
        }
        Ok(self.processed)
    }

    /// Return the number of payments processed.
    pub fn stats(&self) -> u64 {
        self.processed
    }
}
